extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,204 @@
+/// layer and path.
+#[tracing::instrument]
+fn jars_in_container(image_path: &PathBuf) -> Result<HashMap<PathBuf, Vec<DiscoveredJar>>, Error> {
+    // Visit each layer and fingerprint the JARs within.
+    info!("inspecting container");
+    // May have to make two copies of this Archive, idk if it can be iterated twice.
+    let layers = list_container_layers(image_path)?;
+    let mut discoveries = HashMap::new();
+
+    let mut image_archive = unpack(image_path)?;
+    for entry in wrap_tar_err!("Iterate entries", image_archive.entries())? {
+        let entry = wrap_tar_err!("Read entry", entry)?;
+        let path = wrap_tar_err!("Read path", entry.path())?;
+        if !layers.contains(path.as_ref()) {
+            debug!(?path, "skipped: not a layer file");
+            continue;
+        }
+
+        let layer = path.to_path_buf();
+        // Layers should have a form like blob
+        let layer_discoveries = jars_in_layer(entry)?;
+        discoveries.insert(layer, layer_discoveries);
+    }
+
+    Ok(discoveries)
+}
+
+/// Open and unpack a file and put it into a tar.
+/// This is done repeatedly because `entries()` can only be read once from an Archive.
+#[tracing::instrument]
+fn unpack(path: &PathBuf) -> Result<Archive<File>, Error> {
+    let file = File::open(path).map_err(Error::OpenFile)?;
+    Ok(tar::Archive::new(file))
+}
+
+#[tracing::instrument(skip(entry))]
+fn jars_in_layer(entry: Entry<'_, impl Read>) -> Result<Vec<DiscoveredJar>, Error> {
+    let mut discoveries = Vec::new();
+
+    let mut entry_archive = tar::Archive::new(entry);
+    for entry in wrap_tar_err!("list entries in layer", entry_archive.entries())? {
+        let entry = wrap_tar_err!("read entry", entry)?;
+        let path = wrap_tar_err!("read path", entry.path())?;
+        if !path.to_string_lossy().ends_with(".jar") {
+            debug!(?path, "skipped: not a jar file");
+            continue;
+        }
+
+        let path = path.to_path_buf();
+        let res = info_span!("jar", ?path).in_scope(|| {
+            debug!("fingerprinting");
+            let entry = buffer(entry)?;
+
+            match Combined::from_buffer(entry) {
+                Ok(fingerprint) => {
+                    discoveries.push(DiscoveredJar { fingerprint, path });
+                    Ok(())
+                }
+                Err(e) => Err(Error::GenerateFingerprint(e)),
+            }
+        });
+
+        if let Err(e) = res {
+            warn!("Fingerprinting: {:?}", e);
+        }
+    }
+
+    Ok(discoveries)
+}
+
+#[tracing::instrument]
+fn list_container_layers(layer_path: &PathBuf) -> Result<HashSet<PathBuf>, Error> {
+    let mut layers = HashSet::new();
+
+    let mut container = unpack(layer_path)?;
+    for entry in wrap_tar_err!("list entries", container.entries())? {
+        let entry = match entry {
+            Ok(entry) => entry,
+            Err(e) => {
+                warn!("Reading entry {:?}", e);
```

Comment:
Warnings are emitted to users, and unlike errors aren't nested, so make sure they're relatively complete sentences. Also, prefer to inline variables in format statements.
```suggestion
                warn!("failed to read entry: {e:?}");
```